### PR TITLE
Add Kafka runbook

### PR DESCRIPTION
**Pull request by hemajv**
Runbook for Kafka

(this addresses operate-first/support#839)

### CODE REVIEW

File: runbooks/kafka.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+## Issue Escalation?
+
+If we need to escalate an issue with the <> team, instructions for doing so can be found at .
+
+## Kafka Topology
+
+There are several different types of pods associated with our Kafka deployment.
+They are as follows:
+
+| Node Type | Description |
```
Comment (tumido) [inline runbooks/kafka.md]: `pod != node != deployment` in Kubernetes. I think we need to keep the terminology consistent. It should be either all called `pods` or `deployments + statefulsets`.

File: runbooks/kafka.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+| --------- | ----------- |
+| Strimzi cluster operator | This is the main Strimzi operator pod. It orchestrates all other pods. |
```
Comment (tumido) [inline runbooks/kafka.md]: It's a deployment and lives in a completely different namespace. It should be noted that this lives in `openshift-operators` namespace.

File: runbooks/kafka.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+| Zookeeper | Zookeeper pods. These are deployed by the Strimzi operator. The Kafka brokers connect to the zookeeper cluster to store important cluster information. |
+| Kafka Brokers | The Kafka broker pods that handle the Kafka message passing and storage. These are deployed and managed by the Strimzi Operator. |
+| Kafka Connect | These pods run kafka connectors. They connect to the Kafka brokers and perform specialized message passing functions. We do not use Kafka Connect for production workflows. These are deployed and managed by the Strimzi Operator. |
+| Kafka Consumer Monitor | This is a custom deployment not managed by Strimzi. We use it to monitor Kafka consumer lag. |
```
Comment (tumido) [inline runbooks/kafka.md]: We don't deploy a consumer monitor

File: runbooks/kafka.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
+# Kafka Runbook
+
+## Key Locations
+
+* GitHub Repository: [kafka](https://github.com/operate-first/apps/tree/master/odh/overlays/moc/zero/kafka)
+* Production Namespace: [opf-kafka](https://console-openshift-console.apps.zero.massopen.cloud/k8s/cluster/projects/opf-kafka)
+
+## Issue Escalation?
+
+If we need to escalate an issue with the <> team, instructions for doing so can be found at .
```
Comment (tumido) [inline runbooks/kafka.md]: Please file an issue for the escalation procedure and link to it in here - we can't leave this blank.

**Review by tumido — CHANGES_REQUESTED** [review]

File: runbooks/kafka.md (hunk 4 above)
Comment (HumairAK) [reply runbooks/kafka.md]: Hrmm we don't really have someone we can escalate kafka issues to. This runbook is for us, so the only "escalation" we can do is just reach out to someone else to come save us. I'm down to just get rid of escalation, and instead replace it with "People to contact for additional help." I'm thinking things like, strimzi team, Maulik from idh, etc. Not sure -- wdyt?

File: runbooks/kafka.md (hunk 2 above)
Comment (HumairAK) [reply runbooks/kafka.md]: I think we can go a step further and just give the links to these locations to these pods/sts: 
e.g: 

| [Strimzi cluster operator][1] |  `<description>` 
| [Zookeeper][2]| `<description>` 

[1]: https://console-openshift-console.apps.zero.massopen.cloud/k8s/ns/openshift-operators/pods/strimzi-cluster-operator-v0.19.0-7c78d4cc-m654p
[2]: https://console-openshift-console.apps.zero.massopen.cloud/k8s/ns/opf-kafka/statefulsets/odh-message-bus-zookeeper

and note that you need cluster admin to access `openshift-operators` and need to be in the `operate-first` ocp group to access `opf-kafka` namespace.

File: runbooks/kafka.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+Some items to take note of:
+
+| Metric | Purpose | Desired Value | Action if out of desired range |
+| ------ | ------- | ------------- | ------------------------------ |
+| Brokers online | The number of kafka brokers active in the cluster | This should match whatever number is configured in Git for the cluster size (currently 5) | Check kafka broker pods for restarts or pod logs for signs of error |
+| Under replicated partitions | The number of kafka partitions that don't have replication fully in sync | 0 | Check kafka broker pod logs for signs of errors |
+| Offline partitions count | The number of kafka partitions that aren't available | 0 | Check for offline Kafka brokers. Check for kafka topics that don't have replication fully enabled. |
+| Kafka consumer group lag | The delay in message consumption for a particular consumer group | Close to zero, or not rising | Check consumers for a specific consumer group (typically [kafka2es](KAFKA2ES.md) to see why it's not keeping up with the rate of messages. |
```
Comment (HumairAK) [inline runbooks/kafka.md]: do we have a panel for this?? I don't see it ;'( 

If not, let's take it out?

**Comment by HumairAK** [comment]
@maulikjs  -- wondering if you could give this a look over? it needs to be generalized for our kafka deployment via odh: 

https://console-openshift-console.apps.zero.massopen.cloud/k8s/ns/opf-kafka/pods

File: runbooks/kafka.md (hunk 4 above)
Comment (hemajv) [reply runbooks/kafka.md]: sure, I can update that

File: runbooks/kafka.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+## Kafka Topology
+
+There are several different types of pods associated with our Kafka deployment.
+They are as follows:
+
+| Node Type | Description |
+| --------- | ----------- |
+| Strimzi cluster operator | This is the main Strimzi operator pod. It orchestrates all other pods. |
+| Zookeeper | Zookeeper pods. These are deployed by the Strimzi operator. The Kafka brokers connect to the zookeeper cluster to store important cluster information. |
```
Comment (maulikjs) [inline runbooks/kafka.md]: I'd say, "zookeper performs controller election, cluster membership and maintain a list of all topics and configurations for the kafka brokers."

File: runbooks/kafka.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+| Kafka Brokers | The Kafka broker pods that handle the Kafka message passing and storage. These are deployed and managed by the Strimzi Operator. |
```
Comment (maulikjs) [inline runbooks/kafka.md]: Kafka brokers handle and store the topic log partitions and services the consumers and producers.

File: runbooks/kafka.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+| Kafka Connect | These pods run kafka connectors. They connect to the Kafka brokers and perform specialized message passing functions. We do not use Kafka Connect for production workflows. These are deployed and managed by the Strimzi Operator. |
```
Comment (maulikjs) [inline runbooks/kafka.md]: Kafka connect utilizes various connectors to connect the Kafka cluster to a variety of data sources and streams

File: runbooks/kafka.md [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+directory. Some particularly useful scrips are as follows:
+
+| Script | Use |
+| ------ | --- |
+| `kafka-topics.sh` | View topics in the cluster, information about a specific topic, create a topic, etc. |
+| `kafka-consumer-groups.sh` | View statistics about a consumer group |
+| `kafka-console-producer.sh` | Manually produce messages onto a topic |
+| `kafka-console-consumer.sh` | Manually consume messages from a topic |
+| `kafka-reassign-partitions.sh` | Manually reassign partitions of a topic to a different broker node |
+
```
Comment (maulikjs) [inline runbooks/kafka.md:64]: ./kafka-log-dirs.sh 
This tool helps to query log directory usage on the specified brokers.

Even this is an important one

File: runbooks/kafka.md [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+```
+/opt/kafka/bin/kafka-consumer-groups.sh --bootstrap-server three-kafka-bootstrap:9092 --list
+```
+
+If no consumer groups are listed, something is wrong.
+
+### Check consumer group lag
+
+```
+/opt/kafka/bin/kafka-consumer-groups.sh --bootstrap-server three-kafka-bootstrap:9092 --descibe --group $NAME
```
Comment (maulikjs) [inline runbooks/kafka.md]: typo s/descibe/describe

File: runbooks/kafka.md [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+Our primary dashboard for monitoring Kafka can be found [here](https://grafana-route-opf-monitoring.apps.zero.massopen.cloud/d/80de0e219205a33a2d87820f1ce335b6805ddbfc/kafka-overview?orgId=1)
+
+Some items to take note of:
+
+| Metric | Purpose | Desired Value | Action if out of desired range |
+| ------ | ------- | ------------- | ------------------------------ |
+| Brokers online | The number of kafka brokers active in the cluster | This should match whatever number is configured in Git for the cluster size (currently 5) | Check kafka broker pods for restarts or pod logs for signs of error |
+| Under replicated partitions | The number of kafka partitions that don't have replication fully in sync | 0 | Check kafka broker pod logs for signs of errors |
+| Offline partitions count | The number of kafka partitions that aren't available | 0 | Check for offline Kafka brokers. Check for kafka topics that don't have replication fully enabled. |
```
Comment (maulikjs) [inline runbooks/kafka.md]: No of kafka partitions that dont have any replicas online

File: runbooks/kafka.md [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+| Kafka consumer group lag | The delay in message consumption for a particular consumer group | Close to zero, or not rising | Check consumers for a specific consumer group (typically [kafka2es](KAFKA2ES.md) to see why it's not keeping up with the rate of messages. |
+
+## Common Debugging Steps
+
+### Topic partition reassignment
```
Comment (maulikjs) [inline runbooks/kafka.md:98]: You will also need to use this command to decrease the replication factor sometimes.

File: runbooks/kafka.md (hunk 4 above)
Comment (hemajv) [reply runbooks/kafka.md]: do we know an email/chat channel for Strimzi team that we could mention?

File: runbooks/kafka.md (hunk 12 above)
Comment (hemajv) [reply runbooks/kafka.md:98]: added changes! thanks for the inputs @maulikjs :smile:

File: runbooks/kafka.md [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+There are several different types of pods associated with our Kafka deployment.
+They are as follows:
+
+| Pods | Description |
+| --------- | ----------- |
+| [Strimzi cluster operator](https://console-openshift-console.apps.zero.massopen.cloud/k8s/ns/openshift-operators/pods/strimzi-cluster-operator-v0.19.0-7c78d4cc-m654p) | This is the main Strimzi operator pod. It orchestrates all other pods and resides in the `openshift-operators` namespace. |
+| [Zookeeper](https://console-openshift-console.apps.zero.massopen.cloud/k8s/ns/opf-kafka/statefulsets/odh-message-bus-zookeeper) | Zookeper performs controller election, cluster membership and maintain a list of all topics and configurations for the kafka brokers. |
+| Kafka Brokers| Kafka brokers handle and store the topic log partitions and services the consumers and producers. These are deployed and managed by the Strimzi Operator. |
+| Kafka Connect] | Kafka connect utilizes various connectors to connect the Kafka cluster to a variety of data sources and streams. These are deployed and managed by the Strimzi Operator. |
```
Comment (4n4nd) [inline runbooks/kafka.md]: stray character `]`

File: runbooks/kafka.md (hunk 12 above)
Comment (maulikjs) [reply runbooks/kafka.md:98]: My pleasure :D

File: runbooks/kafka.md (hunk 4 above)
Comment (maulikjs) [reply runbooks/kafka.md]: https://strimzi.io/community/

File: runbooks/kafka.md [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+
+## Key Locations
+
+| Env                                    | Namespace                                                    | GitHub Repo                        |
+|----------------------------------------|-----------------------------------------------------------------------------------------------------------|-------------------------------|
+| MOC-ZERO Prod                       | [opf-kafka](https://console-openshift-console.apps.zero.massopen.cloud/k8s/cluster/projects/opf-kafka)                                                                            | https://github.com/operate-first/apps/tree/master/odh/overlays/moc/zero/kafka                        |
+
+## Contact for Additional Help
+
+* Strimzi Team -
```
Comment (HumairAK) [inline runbooks/kafka.md]: did you mean to add links here? (assuming so because of the `-`)

Comment (hemajv) [reply runbooks/kafka.md]: yes, do we have an email id or link to a chat channel maybe?

Comment (HumairAK) [reply runbooks/kafka.md]: Looks like @maulikjs 's link above has  chatroom link for slack: https://slack.cncf.io/

Comment (HumairAK) [reply runbooks/kafka.md]: I don't see very many channels in there though...

File: runbooks/kafka.md [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+| Dashboard Description                     | Dashboard URL     |
+| ------------------------------------------- | ------------ |
+| Kafka Overview                       | https://grafana-route-opf-monitoring.apps.zero.massopen.cloud/d/80de0e219205a33a2d87820f1ce335b6805ddbfc/kafka-overview?orgId=1 |                            |
+
+Some items to take note of:
+
+| Metric | Purpose | Desired Value | Action if out of desired range |
+| ------ | ------- | ------------- | ------------------------------ |
+| Brokers online | The number of kafka brokers active in the cluster | This should match whatever number is configured in Git for the cluster size (currently 3) | Check kafka broker pods for restarts or pod logs for signs of error |
+| Under replicated partitions | The number of kafka partitions that don't have replication fully in sync | 0 | Check kafka broker pod logs for signs of errors |
```
Comment (maulikjs) [inline runbooks/kafka.md]: s/replications/replicas

**Comment by maulikjs** [comment]
Other than the 1 comment /lgtm

**Review by tumido — COMMENTED** [review]
/lgtm
Thank you! This is nice :+1:

**Pull request merged**